router-tests/integration_test.go

```diff
@@ -0,0 +1,282 @@
+	flag.Parse()
+	ctx := context.Background()
+	var runner SubgraphsRunner
+	var err error
+	switch *subgraphsMode {
+	case "in-process":
+		runner, err = NewInProcessSubgraphsRunner()
+	case "subprocess":
+		runner, err = NewSubprocessSubgraphsRunner()
+	case "external":
+		runner, err = NewExternalSubgraphsRunner()
+	default:
+		panic(fmt.Errorf("unknown subgraphs mode %q", *subgraphsMode))
+	}
+	if err != nil {
+		panic(err)
+	}
+	// defer this in case we panic, then call it manually before os.Exit()
+	stop := func() {
+		if err := runner.Stop(ctx); err != nil {
+			panic(err)
+		}
+	}
+	defer stop()
+	go func() {
+		err := runner.Start(ctx)
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	const maxRetries = 10
+
+	// Wait until the ports are open
+	for _, port := range runner.Ports() {
+		retries := 0
+		for {
+			_, err := net.Dial("tcp", "127.0.0.1:"+strconv.Itoa(port))
+			if err == nil {
+				break
+			}
+			retries++
+			if retries > maxRetries {
+				panic(fmt.Errorf("could not connect to port %d after %d retries", port, maxRetries))
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+
+	res := m.Run()
+	stop()
+	os.Exit(res)
+}
+
+func randString(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+	return string(b)
+}
+
+type testQuery struct {
+	Name      string
+	Body      string
+	Variables map[string]interface{}
+}
+
+func (t *testQuery) Data() []byte {
+	name := t.Name
+	if name == "" {
+		name = randString(10)
+	}
+	values := map[string]interface{}{
+		"query":         fmt.Sprintf("query %s %s", name, t.Body),
+		"operationName": name,
+	}
+	if len(t.Variables) > 0 {
+		values["variables"] = t.Variables
+	}
+	data, err := json.Marshal(values)
+	if err != nil {
+		panic(err)
+	}
+	return data
+}
+
+func sendQuery(tb testing.TB, server *core.Server, query *testQuery) string {
+	return sendData(tb, server, query.Data())
+}
+
+func sendData(tb testing.TB, server *core.Server, data []byte) string {
+	rr := httptest.NewRecorder()
+
+	req := httptest.NewRequest("POST", "/graphql", bytes.NewBuffer(data))
+	server.Server.Handler.ServeHTTP(rr, req)
+
+	if rr.Code != 200 {
+		tb.Error("unexpected status code", rr.Code)
+	}
+	return rr.Body.String()
+}
+
+func setupServer(tb testing.TB) *core.Server {
+	ctx := context.Background()
+	cfg := config.Config{
+		Graph: config.Graph{
+			Name: "production",
+		},
+	}
+
+	routerConfig, err := core.SerializeConfigFromFile(filepath.Join("testdata", "config.json"))
+	require.Nil(tb, err)
+
+	ec := zap.NewProductionEncoderConfig()
+	ec.EncodeDuration = zapcore.SecondsDurationEncoder
+	ec.TimeKey = "time"
+
+	syncer := zapcore.AddSync(os.Stderr)
+
+	zapLogger := zap.New(zapcore.NewCore(
+		zapcore.NewConsoleEncoder(ec),
+		syncer,
+		zapcore.ErrorLevel,
+	))
+	rs, err := core.NewRouter(
+		core.WithFederatedGraphName(cfg.Graph.Name),
+		core.WithStaticRouterConfig(routerConfig),
+		core.WithModulesConfig(cfg.Modules),
```

Comment:
You don't need that if you have no modules.

Reply:
Thanks! I've deleted it.